{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata Transformation for Submissions to Single Cell Portal (SCP)\n",
    "\n",
    "#### Author: Hannah Kang\n",
    "\n",
    "##### Study Information: \n",
    "The aim of this study is to understand the many types of neurons that carry out information processing in the brain evolved recently by focusing on the retina, the thin film of neurons in the eye that initiates vision. Our approach entails high-throughput single-cell RNA-seq where gene expression is quantified in hundreds of thousands of single retinal neurons, and computational methods are used to process and integrate these datasets. We have recently completed a project that involves generating and integrating single-cell RNA-seq data across an unprecedented 17 species.\n",
    "\n",
    "Date Last Modified: 2/13/2024 (Before 11/18/2023)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Statements\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference data from Professor Shekhar\n",
    "\n",
    "# data = pd.read_csv(\"rgc_metadata_eg.txt\")\n",
    "# data.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata Transformation for Marmoset Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>orig.file</th>\n",
       "      <th>animal</th>\n",
       "      <th>RNA_snn_res.0.5</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>dendro_order</th>\n",
       "      <th>cell_class</th>\n",
       "      <th>integrated_snn_res.1</th>\n",
       "      <th>integrated_snn_res.0.8</th>\n",
       "      <th>all_BC_labels</th>\n",
       "      <th>labels</th>\n",
       "      <th>type</th>\n",
       "      <th>Mammal_OT</th>\n",
       "      <th>Nonmammal_OT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>possorted_genome_bam_EXKMG:AAAGATGAGCCAGGATx</td>\n",
       "      <td>possorted_genome_bam_EXKMG</td>\n",
       "      <td>1730</td>\n",
       "      <td>1194</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>BC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>f2</td>\n",
       "      <td>BC1A</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>possorted_genome_bam_EXKMG:AAAGATGAGGCCCTCAx</td>\n",
       "      <td>possorted_genome_bam_EXKMG</td>\n",
       "      <td>3341</td>\n",
       "      <td>2034</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>BC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>f1</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>possorted_genome_bam_EXKMG:AAACGGGTCTTGTATCx</td>\n",
       "      <td>possorted_genome_bam_EXKMG</td>\n",
       "      <td>2549</td>\n",
       "      <td>1519</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>BC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>f1</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>possorted_genome_bam_EXKMG:AAATGCCCACTAGTACx</td>\n",
       "      <td>possorted_genome_bam_EXKMG</td>\n",
       "      <td>2145</td>\n",
       "      <td>1471</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>BC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>f2</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>possorted_genome_bam_EXKMG:AAACGGGGTCCTAGCGx</td>\n",
       "      <td>possorted_genome_bam_EXKMG</td>\n",
       "      <td>1445</td>\n",
       "      <td>997</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>BC</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>f4</td>\n",
       "      <td>BC4</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12551</th>\n",
       "      <td>possorted_genome_bam_K564P:TTTGGTTGTTGAGTTCx</td>\n",
       "      <td>possorted_genome_bam_K564P</td>\n",
       "      <td>1709</td>\n",
       "      <td>1169</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>BC</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>f5</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12552</th>\n",
       "      <td>possorted_genome_bam_K564P:TTTATGCCATGTTCCCx</td>\n",
       "      <td>possorted_genome_bam_K564P</td>\n",
       "      <td>1910</td>\n",
       "      <td>1351</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>BC</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>f2</td>\n",
       "      <td>BC1A</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12553</th>\n",
       "      <td>possorted_genome_bam_K564P:TTTATGCCAATCACACx</td>\n",
       "      <td>possorted_genome_bam_K564P</td>\n",
       "      <td>2292</td>\n",
       "      <td>1548</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>BC</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>f2</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12554</th>\n",
       "      <td>possorted_genome_bam_K564P:TTTGTCATCTGTTTGTx</td>\n",
       "      <td>possorted_genome_bam_K564P</td>\n",
       "      <td>2963</td>\n",
       "      <td>1794</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>BC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>f1</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12555</th>\n",
       "      <td>possorted_genome_bam_K564P:TTTATGCCACATGTGTx</td>\n",
       "      <td>possorted_genome_bam_K564P</td>\n",
       "      <td>2230</td>\n",
       "      <td>1493</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>BC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>f1</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12556 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Unnamed: 0  \\\n",
       "0      possorted_genome_bam_EXKMG:AAAGATGAGCCAGGATx   \n",
       "1      possorted_genome_bam_EXKMG:AAAGATGAGGCCCTCAx   \n",
       "2      possorted_genome_bam_EXKMG:AAACGGGTCTTGTATCx   \n",
       "3      possorted_genome_bam_EXKMG:AAATGCCCACTAGTACx   \n",
       "4      possorted_genome_bam_EXKMG:AAACGGGGTCCTAGCGx   \n",
       "...                                             ...   \n",
       "12551  possorted_genome_bam_K564P:TTTGGTTGTTGAGTTCx   \n",
       "12552  possorted_genome_bam_K564P:TTTATGCCATGTTCCCx   \n",
       "12553  possorted_genome_bam_K564P:TTTATGCCAATCACACx   \n",
       "12554  possorted_genome_bam_K564P:TTTGTCATCTGTTTGTx   \n",
       "12555  possorted_genome_bam_K564P:TTTATGCCACATGTGTx   \n",
       "\n",
       "                       orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "0      possorted_genome_bam_EXKMG        1730          1194   \n",
       "1      possorted_genome_bam_EXKMG        3341          2034   \n",
       "2      possorted_genome_bam_EXKMG        2549          1519   \n",
       "3      possorted_genome_bam_EXKMG        2145          1471   \n",
       "4      possorted_genome_bam_EXKMG        1445           997   \n",
       "...                           ...         ...           ...   \n",
       "12551  possorted_genome_bam_K564P        1709          1169   \n",
       "12552  possorted_genome_bam_K564P        1910          1351   \n",
       "12553  possorted_genome_bam_K564P        2292          1548   \n",
       "12554  possorted_genome_bam_K564P        2963          1794   \n",
       "12555  possorted_genome_bam_K564P        2230          1493   \n",
       "\n",
       "                   orig.file  animal  RNA_snn_res.0.5  seurat_clusters  \\\n",
       "0           Marmoset1FoveaS1       1                0                2   \n",
       "1           Marmoset1FoveaS1       1                2                1   \n",
       "2           Marmoset1FoveaS1       1                2                1   \n",
       "3           Marmoset1FoveaS1       1                0                2   \n",
       "4           Marmoset1FoveaS1       1                8                4   \n",
       "...                      ...     ...              ...              ...   \n",
       "12551  AdultmarmosetFovea2S2       2                9                5   \n",
       "12552  AdultmarmosetFovea2S2       2                0                2   \n",
       "12553  AdultmarmosetFovea2S2       2                0                2   \n",
       "12554  AdultmarmosetFovea2S2       2                2                1   \n",
       "12555  AdultmarmosetFovea2S2       2                2                1   \n",
       "\n",
       "       dendro_order cell_class  integrated_snn_res.1  integrated_snn_res.0.8  \\\n",
       "0                 0         BC                     1                       1   \n",
       "1                 1         BC                     0                       0   \n",
       "2                 1         BC                     0                       0   \n",
       "3                 0         BC                     1                       1   \n",
       "4                 3         BC                     3                       3   \n",
       "...             ...        ...                   ...                     ...   \n",
       "12551             4         BC                     4                       4   \n",
       "12552             0         BC                     9                       6   \n",
       "12553             0         BC                     1                       1   \n",
       "12554             1         BC                     0                       0   \n",
       "12555             1         BC                     0                       0   \n",
       "\n",
       "       all_BC_labels  labels type Mammal_OT Nonmammal_OT  \n",
       "0                  1       2   f2      BC1A       absent  \n",
       "1                  2       1   f1    absent       absent  \n",
       "2                  2       1   f1    absent       absent  \n",
       "3                  1       2   f2    absent       absent  \n",
       "4                  4       4   f4       BC4       absent  \n",
       "...              ...     ...  ...       ...          ...  \n",
       "12551              5       5   f5    absent       absent  \n",
       "12552              1       2   f2      BC1A       absent  \n",
       "12553              1       2   f2    absent       absent  \n",
       "12554              2       1   f1    absent       absent  \n",
       "12555              2       1   f1    absent       absent  \n",
       "\n",
       "[12556 rows x 17 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"/Users/hannahkang/Desktop/Shek Lab/Original Files/MarmosetFovea_BC_metadata.csv\"\n",
    "\n",
    "metadata_marmoset = pd.read_csv(file_path)\n",
    "metadata_marmoset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>species</th>\n",
       "      <th>species__ontology_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease__ontology_label</th>\n",
       "      <th>organ</th>\n",
       "      <th>organ__ontology_label</th>\n",
       "      <th>library_preparation_protocol</th>\n",
       "      <th>library_preparation_protocol__ontology_label</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TYPE</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EXKMG_AAAGATGAGCCAGGATx</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EXKMG_AAAGATGAGGCCCTCAx</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EXKMG_AAACGGGTCTTGTATCx</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EXKMG_AAATGCCCACTAGTACx</td>\n",
       "      <td>Marmoset1FoveaS1</td>\n",
       "      <td>1</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12552</th>\n",
       "      <td>K564P_TTTGGTTGTTGAGTTCx</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12553</th>\n",
       "      <td>K564P_TTTATGCCATGTTCCCx</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12554</th>\n",
       "      <td>K564P_TTTATGCCAATCACACx</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12555</th>\n",
       "      <td>K564P_TTTGTCATCTGTTTGTx</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12556</th>\n",
       "      <td>K564P_TTTATGCCACATGTGTx</td>\n",
       "      <td>AdultmarmosetFovea2S2</td>\n",
       "      <td>2</td>\n",
       "      <td>NCBITAXON_9483</td>\n",
       "      <td>Callithrix jacchus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12557 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          NAME           biosample_id donor_id  \\\n",
       "0                         TYPE                  group    group   \n",
       "1      EXKMG_AAAGATGAGCCAGGATx       Marmoset1FoveaS1        1   \n",
       "2      EXKMG_AAAGATGAGGCCCTCAx       Marmoset1FoveaS1        1   \n",
       "3      EXKMG_AAACGGGTCTTGTATCx       Marmoset1FoveaS1        1   \n",
       "4      EXKMG_AAATGCCCACTAGTACx       Marmoset1FoveaS1        1   \n",
       "...                        ...                    ...      ...   \n",
       "12552  K564P_TTTGGTTGTTGAGTTCx  AdultmarmosetFovea2S2        2   \n",
       "12553  K564P_TTTATGCCATGTTCCCx  AdultmarmosetFovea2S2        2   \n",
       "12554  K564P_TTTATGCCAATCACACx  AdultmarmosetFovea2S2        2   \n",
       "12555  K564P_TTTGTCATCTGTTTGTx  AdultmarmosetFovea2S2        2   \n",
       "12556  K564P_TTTATGCCACATGTGTx  AdultmarmosetFovea2S2        2   \n",
       "\n",
       "              species species__ontology_label       disease  \\\n",
       "0               group                   group         group   \n",
       "1      NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "2      NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "3      NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "4      NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "...               ...                     ...           ...   \n",
       "12552  NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "12553  NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "12554  NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "12555  NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "12556  NCBITAXON_9483      Callithrix jacchus  PATO_0000461   \n",
       "\n",
       "      disease__ontology_label           organ organ__ontology_label  \\\n",
       "0                       group           group                 group   \n",
       "1                      normal  UBERON_0000966                retina   \n",
       "2                      normal  UBERON_0000966                retina   \n",
       "3                      normal  UBERON_0000966                retina   \n",
       "4                      normal  UBERON_0000966                retina   \n",
       "...                       ...             ...                   ...   \n",
       "12552                  normal  UBERON_0000966                retina   \n",
       "12553                  normal  UBERON_0000966                retina   \n",
       "12554                  normal  UBERON_0000966                retina   \n",
       "12555                  normal  UBERON_0000966                retina   \n",
       "12556                  normal  UBERON_0000966                retina   \n",
       "\n",
       "      library_preparation_protocol  \\\n",
       "0                            group   \n",
       "1                      EFO_0009899   \n",
       "2                      EFO_0009899   \n",
       "3                      EFO_0009899   \n",
       "4                      EFO_0009899   \n",
       "...                            ...   \n",
       "12552                  EFO_0009899   \n",
       "12553                  EFO_0009899   \n",
       "12554                  EFO_0009899   \n",
       "12555                  EFO_0009899   \n",
       "12556                  EFO_0009899   \n",
       "\n",
       "      library_preparation_protocol__ontology_label      sex  \n",
       "0                                            group    group  \n",
       "1                                        10X 3' v2  unknown  \n",
       "2                                        10X 3' v2  unknown  \n",
       "3                                        10X 3' v2  unknown  \n",
       "4                                        10X 3' v2  unknown  \n",
       "...                                            ...      ...  \n",
       "12552                                    10X 3' v2  unknown  \n",
       "12553                                    10X 3' v2  unknown  \n",
       "12554                                    10X 3' v2  unknown  \n",
       "12555                                    10X 3' v2  unknown  \n",
       "12556                                    10X 3' v2  unknown  \n",
       "\n",
       "[12557 rows x 12 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove the prefix from the \"Unnamed: 0\" column\n",
    "metadata_marmoset['Unnamed: 0'] = metadata_marmoset['Unnamed: 0'].str.replace('possorted_genome_bam_', '') \n",
    "\n",
    "metadata_marmoset = metadata_marmoset.rename(columns={'Unnamed: 0': 'NAME'})\n",
    "metadata_marmoset['NAME'] = metadata_marmoset['NAME'].str.replace(':', '_')\n",
    "\n",
    "# Define column values\n",
    "column_values = {\n",
    "    'biosample_id': metadata_marmoset['orig.file'],\n",
    "    'donor_id': metadata_marmoset['animal'],\n",
    "    'species': 'NCBITAXON_9483',\n",
    "    'species__ontology_label': 'Callithrix jacchus',\n",
    "    'disease': 'PATO_0000461',\n",
    "    'disease__ontology_label': 'normal',\n",
    "    'organ': 'UBERON_0000966',\n",
    "    'organ__ontology_label': 'retina',\n",
    "    'library_preparation_protocol': 'EFO_0009899',\n",
    "    'library_preparation_protocol__ontology_label': \"10X 3' v2\",\n",
    "    'sex': 'unknown'\n",
    "}\n",
    "\n",
    "# Insert columns into DataFrame\n",
    "for i, (col_name, col_values) in enumerate(column_values.items(), start=1):\n",
    "    metadata_marmoset.insert(i, col_name, col_values)\n",
    "\n",
    "# Add 'TYPE' row\n",
    "group_row = pd.Series(['TYPE'] + ['group'] * (len(metadata_marmoset.columns) - 1), index=metadata_marmoset.columns)\n",
    "metadata_marmoset = pd.concat([group_row.to_frame().T, metadata_marmoset], ignore_index=True)\n",
    "\n",
    "# Limit DataFrame to the first 12 columns\n",
    "metadata_marmoset = metadata_marmoset.iloc[:, :12]\n",
    "metadata_marmoset\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata Transformation for Mouse Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>orig.file</th>\n",
       "      <th>animal</th>\n",
       "      <th>RNA_snn_res.0.5</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>dendro_order</th>\n",
       "      <th>integrated_snn_res.0.8</th>\n",
       "      <th>integrated_snn_res.0.5</th>\n",
       "      <th>barcode</th>\n",
       "      <th>annotated</th>\n",
       "      <th>type</th>\n",
       "      <th>Mammal_OT</th>\n",
       "      <th>Nonmammal_OT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>possorted_genome_bam_Z0OYT:AAACGGGCAACACGCCx</td>\n",
       "      <td>possorted_genome_bam_Z0OYT</td>\n",
       "      <td>4424.0</td>\n",
       "      <td>2223.0</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>oncBC_CtC57AllOtherR1_AAACGGGCAACACGCC-1</td>\n",
       "      <td>BC5A</td>\n",
       "      <td>BC5A</td>\n",
       "      <td>absent</td>\n",
       "      <td>BC8/9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>possorted_genome_bam_Z0OYT:AAACCTGAGTGTACGGx</td>\n",
       "      <td>possorted_genome_bam_Z0OYT</td>\n",
       "      <td>5860.0</td>\n",
       "      <td>2708.0</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>oncBC_CtC57AllOtherR1_AAACCTGAGTGTACGG-1</td>\n",
       "      <td>BC4</td>\n",
       "      <td>BC4</td>\n",
       "      <td>BC4</td>\n",
       "      <td>BC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>possorted_genome_bam_Z0OYT:AAAGTAGGTTGTTTGGx</td>\n",
       "      <td>possorted_genome_bam_Z0OYT</td>\n",
       "      <td>4458.0</td>\n",
       "      <td>2149.0</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>oncBC_CtC57AllOtherR1_AAAGTAGGTTGTTTGG-1</td>\n",
       "      <td>RBC</td>\n",
       "      <td>RBC</td>\n",
       "      <td>absent</td>\n",
       "      <td>absent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>possorted_genome_bam_Z0OYT:AAAGTAGCATTCCTCGx</td>\n",
       "      <td>possorted_genome_bam_Z0OYT</td>\n",
       "      <td>1841.0</td>\n",
       "      <td>1025.0</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>oncBC_CtC57AllOtherR1_AAAGTAGCATTCCTCG-1</td>\n",
       "      <td>RBC</td>\n",
       "      <td>RBC</td>\n",
       "      <td>absent</td>\n",
       "      <td>RBC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>possorted_genome_bam_Z0OYT:AAATGCCAGCCACTATx</td>\n",
       "      <td>possorted_genome_bam_Z0OYT</td>\n",
       "      <td>1447.0</td>\n",
       "      <td>990.0</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>oncBC_CtC57AllOtherR1_AAATGCCAGCCACTAT-1</td>\n",
       "      <td>BC1B</td>\n",
       "      <td>BC1B</td>\n",
       "      <td>BC2</td>\n",
       "      <td>BC1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8146</th>\n",
       "      <td>Mouse;possorted_genome_bam_H5JEB:TTTACTGTCGGCA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "      <td>BC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8147</th>\n",
       "      <td>Mouse;possorted_genome_bam_H5JEB:TTTGGTTGTTAAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7</td>\n",
       "      <td>BC3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8148</th>\n",
       "      <td>Mouse;possorted_genome_bam_H5JEB:TTTATGCGTCTGA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8</td>\n",
       "      <td>BC5A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8149</th>\n",
       "      <td>Mouse;possorted_genome_bam_H5JEB:TTTCCTCGTTCAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "      <td>BC3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8150</th>\n",
       "      <td>Mouse;possorted_genome_bam_H5JEB:TTTACTGAGTACA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14</td>\n",
       "      <td>BC1A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8151 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Unnamed: 0  \\\n",
       "0          possorted_genome_bam_Z0OYT:AAACGGGCAACACGCCx   \n",
       "1          possorted_genome_bam_Z0OYT:AAACCTGAGTGTACGGx   \n",
       "2          possorted_genome_bam_Z0OYT:AAAGTAGGTTGTTTGGx   \n",
       "3          possorted_genome_bam_Z0OYT:AAAGTAGCATTCCTCGx   \n",
       "4          possorted_genome_bam_Z0OYT:AAATGCCAGCCACTATx   \n",
       "...                                                 ...   \n",
       "8146  Mouse;possorted_genome_bam_H5JEB:TTTACTGTCGGCA...   \n",
       "8147  Mouse;possorted_genome_bam_H5JEB:TTTGGTTGTTAAA...   \n",
       "8148  Mouse;possorted_genome_bam_H5JEB:TTTATGCGTCTGA...   \n",
       "8149  Mouse;possorted_genome_bam_H5JEB:TTTCCTCGTTCAG...   \n",
       "8150  Mouse;possorted_genome_bam_H5JEB:TTTACTGAGTACA...   \n",
       "\n",
       "                      orig.ident  nCount_RNA  nFeature_RNA         orig.file  \\\n",
       "0     possorted_genome_bam_Z0OYT      4424.0        2223.0  CTRLC57AllOther1   \n",
       "1     possorted_genome_bam_Z0OYT      5860.0        2708.0  CTRLC57AllOther1   \n",
       "2     possorted_genome_bam_Z0OYT      4458.0        2149.0  CTRLC57AllOther1   \n",
       "3     possorted_genome_bam_Z0OYT      1841.0        1025.0  CTRLC57AllOther1   \n",
       "4     possorted_genome_bam_Z0OYT      1447.0         990.0  CTRLC57AllOther1   \n",
       "...                          ...         ...           ...               ...   \n",
       "8146                         NaN         NaN           NaN               NaN   \n",
       "8147                         NaN         NaN           NaN               NaN   \n",
       "8148                         NaN         NaN           NaN               NaN   \n",
       "8149                         NaN         NaN           NaN               NaN   \n",
       "8150                         NaN         NaN           NaN               NaN   \n",
       "\n",
       "      animal  RNA_snn_res.0.5  seurat_clusters  dendro_order  \\\n",
       "0        1.0              8.0             10.0           8.0   \n",
       "1        1.0             10.0              8.0          10.0   \n",
       "2        1.0             13.0              1.0          13.0   \n",
       "3        1.0             11.0              1.0          11.0   \n",
       "4        1.0             40.0             15.0          40.0   \n",
       "...      ...              ...              ...           ...   \n",
       "8146     NaN              NaN              NaN           NaN   \n",
       "8147     NaN              NaN              NaN           NaN   \n",
       "8148     NaN              NaN              NaN           NaN   \n",
       "8149     NaN              NaN              NaN           NaN   \n",
       "8150     NaN              NaN              NaN           NaN   \n",
       "\n",
       "      integrated_snn_res.0.8  integrated_snn_res.0.5  \\\n",
       "0                       14.0                    11.0   \n",
       "1                       11.0                     9.0   \n",
       "2                        0.0                     0.0   \n",
       "3                        1.0                     2.0   \n",
       "4                       20.0                    18.0   \n",
       "...                      ...                     ...   \n",
       "8146                     NaN                     NaN   \n",
       "8147                     NaN                     NaN   \n",
       "8148                     NaN                     NaN   \n",
       "8149                     NaN                     NaN   \n",
       "8150                     NaN                     NaN   \n",
       "\n",
       "                                       barcode annotated  type Mammal_OT  \\\n",
       "0     oncBC_CtC57AllOtherR1_AAACGGGCAACACGCC-1      BC5A  BC5A    absent   \n",
       "1     oncBC_CtC57AllOtherR1_AAACCTGAGTGTACGG-1       BC4   BC4       BC4   \n",
       "2     oncBC_CtC57AllOtherR1_AAAGTAGGTTGTTTGG-1       RBC   RBC    absent   \n",
       "3     oncBC_CtC57AllOtherR1_AAAGTAGCATTCCTCG-1       RBC   RBC    absent   \n",
       "4     oncBC_CtC57AllOtherR1_AAATGCCAGCCACTAT-1      BC1B  BC1B       BC2   \n",
       "...                                        ...       ...   ...       ...   \n",
       "8146                                       NaN       NaN   NaN        12   \n",
       "8147                                       NaN       NaN   NaN         7   \n",
       "8148                                       NaN       NaN   NaN         8   \n",
       "8149                                       NaN       NaN   NaN        10   \n",
       "8150                                       NaN       NaN   NaN        14   \n",
       "\n",
       "     Nonmammal_OT  \n",
       "0           BC8/9  \n",
       "1             BC4  \n",
       "2          absent  \n",
       "3             RBC  \n",
       "4            BC1B  \n",
       "...           ...  \n",
       "8146          BC2  \n",
       "8147         BC3B  \n",
       "8148         BC5A  \n",
       "8149         BC3B  \n",
       "8150         BC1A  \n",
       "\n",
       "[8151 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"/Users/hannahkang/Desktop/Shek Lab/Original Files/Mouse_BC_metadata.csv\"\n",
    "\n",
    "metadata_mouse = pd.read_csv(file_path)\n",
    "metadata_mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>species</th>\n",
       "      <th>species__ontology_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease__ontology_label</th>\n",
       "      <th>organ</th>\n",
       "      <th>organ__ontology_label</th>\n",
       "      <th>library_preparation_protocol</th>\n",
       "      <th>library_preparation_protocol__ontology_label</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TYPE</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Z0OYT_AAACGGGCAACACGCCx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Z0OYT_AAACCTGAGTGTACGGx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Z0OYT_AAAGTAGGTTGTTTGGx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Z0OYT_AAAGTAGCATTCCTCGx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8147</th>\n",
       "      <td>H5JEB_TTTACTGTCGGCATCGx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8148</th>\n",
       "      <td>H5JEB_TTTGGTTGTTAAAGACx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8149</th>\n",
       "      <td>H5JEB_TTTATGCGTCTGATTGx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8150</th>\n",
       "      <td>H5JEB_TTTCCTCGTTCAGCGCx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8151</th>\n",
       "      <td>H5JEB_TTTACTGAGTACATGAx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8152 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         NAME      biosample_id donor_id          species  \\\n",
       "0                        TYPE             group    group            group   \n",
       "1     Z0OYT_AAACGGGCAACACGCCx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "2     Z0OYT_AAACCTGAGTGTACGGx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "3     Z0OYT_AAAGTAGGTTGTTTGGx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "4     Z0OYT_AAAGTAGCATTCCTCGx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "...                       ...               ...      ...              ...   \n",
       "8147  H5JEB_TTTACTGTCGGCATCGx               NaN      NaN  NCBITaxon_10090   \n",
       "8148  H5JEB_TTTGGTTGTTAAAGACx               NaN      NaN  NCBITaxon_10090   \n",
       "8149  H5JEB_TTTATGCGTCTGATTGx               NaN      NaN  NCBITaxon_10090   \n",
       "8150  H5JEB_TTTCCTCGTTCAGCGCx               NaN      NaN  NCBITaxon_10090   \n",
       "8151  H5JEB_TTTACTGAGTACATGAx               NaN      NaN  NCBITaxon_10090   \n",
       "\n",
       "     species__ontology_label       disease disease__ontology_label  \\\n",
       "0                      group         group                   group   \n",
       "1               Mus musculus  PATO_0000461                  normal   \n",
       "2               Mus musculus  PATO_0000461                  normal   \n",
       "3               Mus musculus  PATO_0000461                  normal   \n",
       "4               Mus musculus  PATO_0000461                  normal   \n",
       "...                      ...           ...                     ...   \n",
       "8147            Mus musculus  PATO_0000461                  normal   \n",
       "8148            Mus musculus  PATO_0000461                  normal   \n",
       "8149            Mus musculus  PATO_0000461                  normal   \n",
       "8150            Mus musculus  PATO_0000461                  normal   \n",
       "8151            Mus musculus  PATO_0000461                  normal   \n",
       "\n",
       "               organ organ__ontology_label library_preparation_protocol  \\\n",
       "0              group                 group                        group   \n",
       "1     UBERON_0000966                retina                  EFO_0009899   \n",
       "2     UBERON_0000966                retina                  EFO_0009899   \n",
       "3     UBERON_0000966                retina                  EFO_0009899   \n",
       "4     UBERON_0000966                retina                  EFO_0009899   \n",
       "...              ...                   ...                          ...   \n",
       "8147  UBERON_0000966                retina                  EFO_0009899   \n",
       "8148  UBERON_0000966                retina                  EFO_0009899   \n",
       "8149  UBERON_0000966                retina                  EFO_0009899   \n",
       "8150  UBERON_0000966                retina                  EFO_0009899   \n",
       "8151  UBERON_0000966                retina                  EFO_0009899   \n",
       "\n",
       "     library_preparation_protocol__ontology_label      sex  \n",
       "0                                           group    group  \n",
       "1                                       10X 3' v2  unknown  \n",
       "2                                       10X 3' v2  unknown  \n",
       "3                                       10X 3' v2  unknown  \n",
       "4                                       10X 3' v2  unknown  \n",
       "...                                           ...      ...  \n",
       "8147                                    10X 3' v2  unknown  \n",
       "8148                                    10X 3' v2  unknown  \n",
       "8149                                    10X 3' v2  unknown  \n",
       "8150                                    10X 3' v2  unknown  \n",
       "8151                                    10X 3' v2  unknown  \n",
       "\n",
       "[8152 rows x 12 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove the prefix from the \"Unnamed: 0\" column\n",
    "metadata_mouse.loc[:, 'Unnamed: 0'] = metadata_mouse['Unnamed: 0'].str.replace('Mouse;possorted_genome_bam_', '')\n",
    "metadata_mouse.loc[:, 'Unnamed: 0'] = metadata_mouse['Unnamed: 0'].str.replace('possorted_genome_bam_', '')\n",
    "\n",
    "metadata_mouse = metadata_mouse.rename(columns={'Unnamed: 0': 'NAME'})\n",
    "metadata_mouse['NAME'] = metadata_mouse['NAME'].str.replace(':', '_')\n",
    "\n",
    "# Define column values\n",
    "column_values = {\n",
    "    'biosample_id': metadata_mouse['orig.file'],\n",
    "    'donor_id': metadata_mouse['animal'],\n",
    "    'species': 'NCBITaxon_10090',\n",
    "    'species__ontology_label': 'Mus musculus',\n",
    "    'disease': 'PATO_0000461',\n",
    "    'disease__ontology_label': 'normal',\n",
    "    'organ': 'UBERON_0000966',\n",
    "    'organ__ontology_label': 'retina',\n",
    "    'library_preparation_protocol': 'EFO_0009899',\n",
    "    'library_preparation_protocol__ontology_label': \"10X 3' v2\",\n",
    "    'sex': 'unknown'\n",
    "}\n",
    "\n",
    "# Insert columns into DataFrame\n",
    "for i, (col_name, col_values) in enumerate(column_values.items(), start=1):\n",
    "    metadata_mouse.insert(i, col_name, col_values)\n",
    "\n",
    "# Add 'TYPE' row\n",
    "group_row = pd.Series(['TYPE'] + ['group'] * (len(metadata_mouse.columns) - 1), index=metadata_mouse.columns)\n",
    "metadata_mouse = pd.concat([group_row.to_frame().T, metadata_mouse], ignore_index=True)\n",
    "\n",
    "# Limit DataFrame to the first 12 columns\n",
    "metadata_mouse = metadata_mouse.iloc[:, :12]\n",
    "metadata_mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>biosample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>species</th>\n",
       "      <th>species__ontology_label</th>\n",
       "      <th>disease</th>\n",
       "      <th>disease__ontology_label</th>\n",
       "      <th>organ</th>\n",
       "      <th>organ__ontology_label</th>\n",
       "      <th>library_preparation_protocol</th>\n",
       "      <th>library_preparation_protocol__ontology_label</th>\n",
       "      <th>sex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TYPE</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "      <td>group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Z0OYT_AAACGGGCAACACGCCx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Z0OYT_AAACCTGAGTGTACGGx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Z0OYT_AAAGTAGGTTGTTTGGx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Z0OYT_AAAGTAGCATTCCTCGx</td>\n",
       "      <td>CTRLC57AllOther1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5551</th>\n",
       "      <td>H5JEB_TTTACTGTCCGCTGTTx</td>\n",
       "      <td>CTRLC57AllOther2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5552</th>\n",
       "      <td>H5JEB_TTTATGCGTCTGATTGx</td>\n",
       "      <td>CTRLC57AllOther2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5553</th>\n",
       "      <td>H5JEB_TTTCCTCTCTTGCAAGx</td>\n",
       "      <td>CTRLC57AllOther2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5554</th>\n",
       "      <td>H5JEB_TTTCCTCGTTCAGCGCx</td>\n",
       "      <td>CTRLC57AllOther2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5555</th>\n",
       "      <td>H5JEB_TTTACTGAGTACATGAx</td>\n",
       "      <td>CTRLC57AllOther2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NCBITaxon_10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>PATO_0000461</td>\n",
       "      <td>normal</td>\n",
       "      <td>UBERON_0000966</td>\n",
       "      <td>retina</td>\n",
       "      <td>EFO_0009899</td>\n",
       "      <td>10X 3' v2</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5556 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         NAME      biosample_id donor_id          species  \\\n",
       "0                        TYPE             group    group            group   \n",
       "1     Z0OYT_AAACGGGCAACACGCCx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "2     Z0OYT_AAACCTGAGTGTACGGx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "3     Z0OYT_AAAGTAGGTTGTTTGGx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "4     Z0OYT_AAAGTAGCATTCCTCGx  CTRLC57AllOther1      1.0  NCBITaxon_10090   \n",
       "...                       ...               ...      ...              ...   \n",
       "5551  H5JEB_TTTACTGTCCGCTGTTx  CTRLC57AllOther2      1.0  NCBITaxon_10090   \n",
       "5552  H5JEB_TTTATGCGTCTGATTGx  CTRLC57AllOther2      1.0  NCBITaxon_10090   \n",
       "5553  H5JEB_TTTCCTCTCTTGCAAGx  CTRLC57AllOther2      1.0  NCBITaxon_10090   \n",
       "5554  H5JEB_TTTCCTCGTTCAGCGCx  CTRLC57AllOther2      1.0  NCBITaxon_10090   \n",
       "5555  H5JEB_TTTACTGAGTACATGAx  CTRLC57AllOther2      1.0  NCBITaxon_10090   \n",
       "\n",
       "     species__ontology_label       disease disease__ontology_label  \\\n",
       "0                      group         group                   group   \n",
       "1               Mus musculus  PATO_0000461                  normal   \n",
       "2               Mus musculus  PATO_0000461                  normal   \n",
       "3               Mus musculus  PATO_0000461                  normal   \n",
       "4               Mus musculus  PATO_0000461                  normal   \n",
       "...                      ...           ...                     ...   \n",
       "5551            Mus musculus  PATO_0000461                  normal   \n",
       "5552            Mus musculus  PATO_0000461                  normal   \n",
       "5553            Mus musculus  PATO_0000461                  normal   \n",
       "5554            Mus musculus  PATO_0000461                  normal   \n",
       "5555            Mus musculus  PATO_0000461                  normal   \n",
       "\n",
       "               organ organ__ontology_label library_preparation_protocol  \\\n",
       "0              group                 group                        group   \n",
       "1     UBERON_0000966                retina                  EFO_0009899   \n",
       "2     UBERON_0000966                retina                  EFO_0009899   \n",
       "3     UBERON_0000966                retina                  EFO_0009899   \n",
       "4     UBERON_0000966                retina                  EFO_0009899   \n",
       "...              ...                   ...                          ...   \n",
       "5551  UBERON_0000966                retina                  EFO_0009899   \n",
       "5552  UBERON_0000966                retina                  EFO_0009899   \n",
       "5553  UBERON_0000966                retina                  EFO_0009899   \n",
       "5554  UBERON_0000966                retina                  EFO_0009899   \n",
       "5555  UBERON_0000966                retina                  EFO_0009899   \n",
       "\n",
       "     library_preparation_protocol__ontology_label      sex  \n",
       "0                                           group    group  \n",
       "1                                       10X 3' v2  unknown  \n",
       "2                                       10X 3' v2  unknown  \n",
       "3                                       10X 3' v2  unknown  \n",
       "4                                       10X 3' v2  unknown  \n",
       "...                                           ...      ...  \n",
       "5551                                    10X 3' v2  unknown  \n",
       "5552                                    10X 3' v2  unknown  \n",
       "5553                                    10X 3' v2  unknown  \n",
       "5554                                    10X 3' v2  unknown  \n",
       "5555                                    10X 3' v2  unknown  \n",
       "\n",
       "[5556 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_mouse = metadata_mouse.drop_duplicates(subset='NAME', keep='first')\n",
    "metadata_mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Error Checking for Transformation for Species Metadata and UMAP (Clustering) Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NAME</th>\n",
       "      <th>X</th>\n",
       "      <th>Y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TYPE</td>\n",
       "      <td>numeric</td>\n",
       "      <td>numeric</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Z0OYT_AAACGGGCAACACGCCx</td>\n",
       "      <td>12.2872521763422</td>\n",
       "      <td>10.8380569849402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Z0OYT_AAACCTGAGTGTACGGx</td>\n",
       "      <td>6.87598775678653</td>\n",
       "      <td>1.99284834007327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Z0OYT_AAAGTAGGTTGTTTGGx</td>\n",
       "      <td>-9.62521959489804</td>\n",
       "      <td>3.70287269691531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Z0OYT_AAAGTAGCATTCCTCGx</td>\n",
       "      <td>-7.14581848329525</td>\n",
       "      <td>5.00803083519045</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      NAME                  X                 Y\n",
       "0                     TYPE            numeric           numeric\n",
       "1  Z0OYT_AAACGGGCAACACGCCx   12.2872521763422  10.8380569849402\n",
       "2  Z0OYT_AAACCTGAGTGTACGGx   6.87598775678653  1.99284834007327\n",
       "3  Z0OYT_AAAGTAGGTTGTTTGGx  -9.62521959489804  3.70287269691531\n",
       "4  Z0OYT_AAAGTAGCATTCCTCGx  -7.14581848329525  5.00803083519045"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"/Users/hannahkang/Desktop/Shek Lab/Single Cell Portal/Output Files/Marmoset/SCP-marmoset-clustering-submission.csv\"\n",
    "clustering_marmoset = pd.read_csv(file_path)\n",
    "\n",
    "file_path = \"/Users/hannahkang/Desktop/Shek Lab/Single Cell Portal/Output Files/Mouse/SCP-mouse-clustering-submission.csv\"\n",
    "clustering_mouse = pd.read_csv(file_path)\n",
    "\n",
    "# Example of Clustering File Format\n",
    "clustering_mouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALL PASSED!\n"
     ]
    }
   ],
   "source": [
    "# OBJECTIVE: Check for any cells present within either a cluster or metadata file for a species that do not exist in the other file. \n",
    "# In order to be correct, the set() for all test cases should be empty. \n",
    "\n",
    "# FOR MARMOSETS:\n",
    "cell_names_metadata = set(metadata_marmoset['NAME'])\n",
    "cell_names_cluster = set(clustering_marmoset['NAME'])\n",
    "\n",
    "cells_not_in_cluster = cell_names_metadata - cell_names_cluster\n",
    "cells_not_in_metadata = cell_names_cluster - cell_names_metadata\n",
    "\n",
    "# print(\"Cells present in metadata_marmoset but not in cluster_file:\", cells_not_in_cluster)\n",
    "# print(\"Cells present in cluster_file but not in metadata_marmoset:\", cells_not_in_metadata)\n",
    "\n",
    "# FOR MOUSE:\n",
    "cell_names_metadata = set(metadata_mouse['NAME'])\n",
    "cell_names_cluster = set(clustering_mouse['NAME'])\n",
    "\n",
    "cells_not_in_cluster = cell_names_metadata - cell_names_cluster\n",
    "cells_not_in_metadata = cell_names_cluster - cell_names_metadata\n",
    "\n",
    "# print(\"Cells present in metadata_mouse but not in cluster_file:\", cells_not_in_cluster)\n",
    "# print(\"Cells present in cluster_file but not in metadata_mouse:\", cells_not_in_metadata)\n",
    "\n",
    "if (cells_not_in_cluster == set()) and (cells_not_in_metadata == set()):\n",
    "    print(\"ALL PASSED!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CellxGene h5ad File Conversion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# ann_data_2 = sc.read_h5ad(\"/Users/hannahkang/Desktop/Shek Lab/CellxGene/MouseBC.h5ad\", backed = 'r+')\n",
    "# ann_data_2.write_h5ad('mouseBC_modified2.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouseBC = sc.read_h5ad(\"/Users/hannahkang/Desktop/Shek Lab/Original Files/MouseBC.h5ad\", backed = 'r+')\n",
    "mouseBC.__dict__['_raw'].__dict__['_var'] = mouseBC.__dict__['_raw'].__dict__['_var'].rename(columns={'_index': 'features'})\n",
    "# mouseBC.write_h5ad(\"MouseBC_v2.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AxisArrays with keys: X_tsne, X_umap\n",
      "[[ 12.28725218  10.83805698]\n",
      " [  6.87598776   1.99284834]\n",
      " [ -9.62521959   3.7028727 ]\n",
      " ...\n",
      " [-10.01020552   3.79835219]\n",
      " [ 11.4269874   -2.27397066]\n",
      " [ 16.8380134    5.4951424 ]]\n",
      " \n"
     ]
    }
   ],
   "source": [
    "print(mouseBC.obsm)\n",
    "print(mouseBC.obsm[\"X_umap\"])\n",
    "\n",
    "print(' ')\n",
    "#print(ann_data_marmoset.obsm)\n",
    "#print(ann_data_marmoset.obsm[\"X_umap\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouseBC.uns = {\"Title\": \"Mouse Bipolar\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method AnnData.obs_keys of AnnData object with n_obs × n_vars = 5555 × 31053 backed at '/Users/hannahkang/Desktop/Shek Lab/Original Files/MouseBC.h5ad'\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'orig.file', 'animal', 'RNA_snn_res.0.5', 'seurat_clusters', 'dendro_order', 'integrated_snn_res.0.8', 'integrated_snn_res.0.5', 'barcode', 'annotated', 'type'\n",
      "    var: 'features'\n",
      "    uns: 'Title'\n",
      "    obsm: 'X_tsne', 'X_umap'>\n"
     ]
    }
   ],
   "source": [
    "print(mouseBC.obs_keys)\n",
    "\n",
    "mouseBC.obs[\"organism_ontology_term_id\"] = \"NCBITaxon:10090\"\n",
    "mouseBC.obs[\"tissue_ontology_term_id\"] = \"UBERON_0000966\"\n",
    "mouseBC.obs[\"assay_ontology_term_id\"] = \"EFO:0009809\"\n",
    "mouseBC.obs[\"disease_ontology_term_id\"] = \"PATO_0000461\"\n",
    "mouseBC.obs[\"cell_type_ontology_term_id\"] = \"UBERON_0000966\"\n",
    "mouseBC.obs[\"self_reported_ethnicity_ontology_term_id\"] = \"NA\"\n",
    "mouseBC.obs[\"development_stage_ontology_term_id\"] = \"MmusDv\"\n",
    "mouseBC.obs[\"sex_ontology_term_id\"] = \"Unknown\"\n",
    "mouseBC.obs[\"donor_id\"] = \"P56_Batch1\"\n",
    "mouseBC.obs[\"suspension_type\"] = \"cell\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouseBC.write_h5ad('/Users/hannahkang/Desktop/Shek Lab/CellxGene/Output Files/Mouse/mouseBC_modified_real.h5ad')\n",
    "\n",
    "# ann_data_3 = sc.read_h5ad(\"mouseBC_modified_real.h5ad\", backed = 'r+')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "XKR4       False\n",
       "GM37381     True\n",
       "RP1        False\n",
       "SOX17      False\n",
       "GM37323     True\n",
       "           ...  \n",
       "GM28672     True\n",
       "GM28670     True\n",
       "GM29504     True\n",
       "GM20837     True\n",
       "GM47283     True\n",
       "Name: gm, Length: 31053, dtype: bool"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#returns a boolean for each gene if it starts with the letters 'GM' or not\n",
    "mouseBC.var['gm'] = mouseBC.var_names.str.startswith('GM')\n",
    "mouseBC.var['gm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(mouseBC)\n",
    "# sc.pp.calculate_qc_metrics(mouseBC, qc_vars=['gm'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AxisArrays with keys: X_tsne, X_umap\n",
      "[[ 12.28725218  10.83805698]\n",
      " [  6.87598776   1.99284834]\n",
      " [ -9.62521959   3.7028727 ]\n",
      " ...\n",
      " [-10.01020552   3.79835219]\n",
      " [ 11.4269874   -2.27397066]\n",
      " [ 16.8380134    5.4951424 ]]\n"
     ]
    }
   ],
   "source": [
    "#print(ann_data.X[:10])\n",
    "#print(ann_data.raw.X[:10])\n",
    "print(mouseBC.obsm)\n",
    "print(mouseBC.obsm[\"X_umap\"])\n",
    "\n",
    "# print(' ')\n",
    "# print(ann_data_marmoset.obsm)\n",
    "# print(ann_data_marmoset.obsm[\"X_umap\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['XKR4', 'GM37381', 'RP1', 'SOX17', 'GM37323', 'MRPL15', 'RGS20',\n",
       "       'NPBWR1', '4732440D04RIK', 'GM26901',\n",
       "       ...\n",
       "       'GM28406', 'GM29436', 'GM28407', 'GM29393', 'GM21294', 'GM28672',\n",
       "       'GM28670', 'GM29504', 'GM20837', 'GM47283'],\n",
       "      dtype='object', length=31053)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Fetching the variable names of the h5ad data\n",
    "mouseBC.var_names\n",
    "# ann_data_marmoset.var_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "mouseBC.var_names_make_unique()\n",
    "# ann_data_marmoset.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['possorted_genome_bam_Z0OYT:AAACGGGCAACACGCCx',\n",
       "       'possorted_genome_bam_Z0OYT:AAACCTGAGTGTACGGx',\n",
       "       'possorted_genome_bam_Z0OYT:AAAGTAGGTTGTTTGGx',\n",
       "       'possorted_genome_bam_Z0OYT:AAAGTAGCATTCCTCGx',\n",
       "       'possorted_genome_bam_Z0OYT:AAATGCCAGCCACTATx',\n",
       "       'possorted_genome_bam_Z0OYT:AAACGGGTCTTGCCGTx',\n",
       "       'possorted_genome_bam_Z0OYT:AAAGCAATCAGTACGTx',\n",
       "       'possorted_genome_bam_Z0OYT:AAACGGGTCAGCAACTx',\n",
       "       'possorted_genome_bam_Z0OYT:AAACGGGGTAGGCATGx',\n",
       "       'possorted_genome_bam_Z0OYT:AAAGATGTCTCGATGAx',\n",
       "       ...\n",
       "       'possorted_genome_bam_H5JEB:TTTCCTCAGGCTATCTx',\n",
       "       'possorted_genome_bam_H5JEB:TTTGGTTTCTCCGGTTx',\n",
       "       'possorted_genome_bam_H5JEB:TTTCCTCAGGTAAACTx',\n",
       "       'possorted_genome_bam_H5JEB:TTTACTGTCGGCATCGx',\n",
       "       'possorted_genome_bam_H5JEB:TTTGGTTGTTAAAGACx',\n",
       "       'possorted_genome_bam_H5JEB:TTTACTGTCCGCTGTTx',\n",
       "       'possorted_genome_bam_H5JEB:TTTATGCGTCTGATTGx',\n",
       "       'possorted_genome_bam_H5JEB:TTTCCTCTCTTGCAAGx',\n",
       "       'possorted_genome_bam_H5JEB:TTTCCTCGTTCAGCGCx',\n",
       "       'possorted_genome_bam_H5JEB:TTTACTGAGTACATGAx'],\n",
       "      dtype='object', length=5555)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouseBC.obs_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CSRDataset: backend hdf5, shape (5555, 31053), data_dtype float64\n"
     ]
    }
   ],
   "source": [
    "print(mouseBC.X)\n",
    "\n",
    "# print(ann_data_marmoset.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#returns a boolean for each gene if it starts with the letters 'GM' or not\n",
    "# mouseBC.var['gm'] = mouseBC.var_names.str.startswith('GM')\n",
    "# mouseBC.var['gm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method AnnData.obs_keys of AnnData object with n_obs × n_vars = 5555 × 31053 backed at '/Users/hannahkang/Desktop/Shek Lab/CellxGene/Output Files/Mouse/mouseBC_modified_real.h5ad'\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'orig.file', 'animal', 'RNA_snn_res.0.5', 'seurat_clusters', 'dendro_order', 'integrated_snn_res.0.8', 'integrated_snn_res.0.5', 'barcode', 'annotated', 'type', 'organism_ontology_term_id', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'donor_id', 'suspension_type'\n",
      "    var: 'features', 'gm'\n",
      "    uns: 'Title'\n",
      "    obsm: 'X_tsne', 'X_umap'>\n",
      "possorted_genome_bam_Z0OYT:AAACGGGCAACACGCCx    NA\n",
      "possorted_genome_bam_Z0OYT:AAACCTGAGTGTACGGx    NA\n",
      "possorted_genome_bam_Z0OYT:AAAGTAGGTTGTTTGGx    NA\n",
      "possorted_genome_bam_Z0OYT:AAAGTAGCATTCCTCGx    NA\n",
      "possorted_genome_bam_Z0OYT:AAATGCCAGCCACTATx    NA\n",
      "                                                ..\n",
      "possorted_genome_bam_H5JEB:TTTACTGTCCGCTGTTx    NA\n",
      "possorted_genome_bam_H5JEB:TTTATGCGTCTGATTGx    NA\n",
      "possorted_genome_bam_H5JEB:TTTCCTCTCTTGCAAGx    NA\n",
      "possorted_genome_bam_H5JEB:TTTCCTCGTTCAGCGCx    NA\n",
      "possorted_genome_bam_H5JEB:TTTACTGAGTACATGAx    NA\n",
      "Name: self_reported_ethnicity_ontology_term_id, Length: 5555, dtype: category\n",
      "Categories (1, object): ['NA']\n"
     ]
    }
   ],
   "source": [
    "# print(mouseBC)\n",
    "# sc.pp.calculate_qc_metrics(mouseBC, qc_vars=['gm'], inplace=True)\n",
    "\n",
    "print(mouseBC.obs_keys)\n",
    "\n",
    "# mouseBC.obs[\"organism_ontology_term_id\"] = \"NCBITaxon:10090\"\n",
    "# mouseBC.obs[\"tissue_ontology_term_id\"] = \"UBERON_0000966\"\n",
    "# mouseBC.obs[\"assay_ontology_term_id\"] = \"EFO:0009809\"\n",
    "# mouseBC.obs[\"disease_ontology_term_id\"] = \"PATO_0000461\"\n",
    "# mouseBC.obs[\"cell_type_ontology_term_id\"] = \"UBERON_0000966\"\n",
    "# mouseBC.obs[\"self_reported_ethnicity_ontology_term_id\"] = \"NA\"\n",
    "# mouseBC.obs[\"development_stage_ontology_term_id\"] = \"MmusDv\"\n",
    "# mouseBC.obs[\"sex_ontology_term_id\"] = \"Unknown\"\n",
    "# mouseBC.obs[\"donor_id\"] = \"P56_Batch1\"\n",
    "# mouseBC.obs[\"suspension_type\"] = \"cell\"\n",
    "\n",
    "print(mouseBC.obs[\"self_reported_ethnicity_ontology_term_id\"])\n",
    "\n",
    "#Save the modified ann_data file\n",
    "# mouseBC.write_h5ad('mouseBC_modified.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(ann_data_marmoset.obsm[\"X_umap\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5555 × 31053\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'orig.file', 'animal', 'RNA_snn_res.0.5', 'seurat_clusters', 'dendro_order', 'integrated_snn_res.0.8', 'integrated_snn_res.0.5', 'barcode', 'annotated', 'type', 'organism_ontology_term_id', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'donor_id', 'suspension_type'\n",
       "    var: 'features'\n",
       "    uns: 'Title'\n",
       "    obsm: 'X_tsne', 'X_umap'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dario's reading the file\n",
    "\n",
    "mouseBC = sc.read_h5ad(\"/Users/hannahkang/Desktop/Shek Lab/CellxGene/Output Files/Mouse/mouseBC_modified_real.h5ad\")\n",
    "mouseBC\n",
    "\n",
    "# Example of output: \n",
    "# AnnData object with n_obs × n_vars = 5555 × 31053\n",
    "#     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'orig.file', 'animal', 'RNA_snn_res.0.5', 'seurat_clusters', 'dendro_order', 'integrated_snn_res.0.8', 'integrated_snn_res.0.5', 'barcode', 'annotated', 'type', 'organism_ontology_term_id', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'donor_id', 'suspension_type'\n",
    "#     var: 'features'\n",
    "#     uns: 'Title'\n",
    "#     obsm: 'X_tsne', 'X_umap'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AxisArrays with keys: X_tsne, X_umap\n",
      "[[  0.603162     9.92028538]\n",
      " [-10.56769328   0.07132263]\n",
      " [-10.37739902   1.68855516]\n",
      " ...\n",
      " [  0.98721249   9.85885254]\n",
      " [-10.08541541   1.2832434 ]\n",
      " [-10.49356132   1.53761617]]\n",
      "<bound method AnnData.obs_keys of AnnData object with n_obs × n_vars = 12556 × 27665 backed at '/Users/hannahkang/Desktop/Shek Lab/Original Files/MarmosetFoveaBC.h5ad'\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'orig.file', 'animal', 'RNA_snn_res.0.5', 'seurat_clusters', 'dendro_order', 'cell_class', 'integrated_snn_res.1', 'integrated_snn_res.0.8', 'all_BC_labels', 'labels', 'type'\n",
      "    var: 'features'\n",
      "    uns: 'Title'\n",
      "    obsm: 'X_tsne', 'X_umap'>\n",
      "AxisArrays with keys: X_tsne, X_umap\n",
      "[[  0.603162     9.92028538]\n",
      " [-10.56769328   0.07132263]\n",
      " [-10.37739902   1.68855516]\n",
      " ...\n",
      " [  0.98721249   9.85885254]\n",
      " [-10.08541541   1.2832434 ]\n",
      " [-10.49356132   1.53761617]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 12556 × 27665\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'orig.file', 'animal', 'RNA_snn_res.0.5', 'seurat_clusters', 'dendro_order', 'cell_class', 'integrated_snn_res.1', 'integrated_snn_res.0.8', 'all_BC_labels', 'labels', 'type', 'organism_ontology_term_id', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'donor_id', 'suspension_type'\n",
       "    var: 'features'\n",
       "    uns: 'Title'\n",
       "    obsm: 'X_tsne', 'X_umap'"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Marmoset\n",
    "\n",
    "ann_data_marmoset = sc.read_h5ad(\"/Users/hannahkang/Desktop/Shek Lab/Original Files/MarmosetFoveaBC.h5ad\", backed = 'r+')\n",
    "ann_data_marmoset.__dict__['_raw'].__dict__['_var'] = ann_data_marmoset.__dict__['_raw'].__dict__['_var'].rename(columns={'_index': 'features'})\n",
    "# ann_data_marmoset.write_h5ad(\"MarmosetBC_v2.h5ad\")\n",
    "\n",
    "print(ann_data_marmoset.obsm)\n",
    "print(ann_data_marmoset.obsm[\"X_umap\"])\n",
    "\n",
    "ann_data_marmoset.uns = {\"Title\": \"Marmoset Bipolar\"}\n",
    "\n",
    "print(ann_data_marmoset.obs_keys)\n",
    "\n",
    "# MUST CHECK IF THE DATA IS ACTUALLY CORRECT !!!\n",
    "ann_data_marmoset.obs[\"organism_ontology_term_id\"] = \"NCBITaxon:10090\"\n",
    "ann_data_marmoset.obs[\"tissue_ontology_term_id\"] = \"UBERON_0000966\"\n",
    "ann_data_marmoset.obs[\"assay_ontology_term_id\"] = \"EFO:0009809\"\n",
    "ann_data_marmoset.obs[\"disease_ontology_term_id\"] = \"PATO_0000461\"\n",
    "ann_data_marmoset.obs[\"cell_type_ontology_term_id\"] = \"UBERON_0000966\"\n",
    "ann_data_marmoset.obs[\"self_reported_ethnicity_ontology_term_id\"] = \"NA\"\n",
    "ann_data_marmoset.obs[\"development_stage_ontology_term_id\"] = \"MmusDv\"\n",
    "ann_data_marmoset.obs[\"sex_ontology_term_id\"] = \"Unknown\"\n",
    "ann_data_marmoset.obs[\"donor_id\"] = \"P56_Batch1\"\n",
    "ann_data_marmoset.obs[\"suspension_type\"] = \"cell\"\n",
    "\n",
    "ann_data_marmoset.write_h5ad('/Users/hannahkang/Desktop/Shek Lab/CellxGene/Output Files/Marmoset/ann_data_marmoset_modified_real.h5ad')\n",
    "\n",
    "# returns a boolean for each gene if it starts with the letters 'GM' or not\n",
    "ann_data_marmoset.var['gm'] = ann_data_marmoset.var_names.str.startswith('GM')\n",
    "ann_data_marmoset.var['gm']\n",
    "\n",
    "print(ann_data_marmoset.obsm)\n",
    "print(ann_data_marmoset.obsm[\"X_umap\"])\n",
    "\n",
    "#Fetching the variable names of the h5ad data\n",
    "ann_data_marmoset.var_names\n",
    "\n",
    "ann_data_marmoset.var_names_make_unique()\n",
    "\n",
    "ann_data_marmoset.obs_names\n",
    "\n",
    "# print(ann_data_marmoset.X)\n",
    "\n",
    "# Dario's reading the file\n",
    "ann_data_marmoset = sc.read_h5ad(\"/Users/hannahkang/Desktop/Shek Lab/CellxGene/Output Files/Marmoset/ann_data_marmoset_modified_real.h5ad\")\n",
    "ann_data_marmoset\n",
    "# Example of output: \n",
    "# AnnData object with n_obs × n_vars = 5555 × 31053\n",
    "#     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'orig.file', 'animal', 'RNA_snn_res.0.5', 'seurat_clusters', 'dendro_order', 'integrated_snn_res.0.8', 'integrated_snn_res.0.5', 'barcode', 'annotated', 'type', 'organism_ontology_term_id', 'tissue_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'cell_type_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'donor_id', 'suspension_type'\n",
    "#     var: 'features'\n",
    "#     uns: 'Title'\n",
    "#     obsm: 'X_tsne', 'X_umap'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Translate Dataframes to CSV Files for Submission to SCP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = \"/Users/hannahkang/Desktop/Shek Lab/Single Cell Portal/Output Files/Marmoset/SCP-metadata-submission-marmoset.csv\"\n",
    "metadata_marmoset.to_csv(file_path, index=False)\n",
    "\n",
    "file_path = \"/Users/hannahkang/Desktop/Shek Lab/Single Cell Portal/Output Files/Mouse/SCP-metadata-submission-mouse.csv\"\n",
    "metadata_mouse.to_csv(file_path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scratch Work Code (Ignore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# metadata_marmoset.insert(1, 'biosample_id', metadata_marmoset['orig.file'])\n",
    "# metadata_marmoset.insert(2, 'donor_id', metadata_marmoset['animal'])\n",
    "# metadata_marmoset.insert(3, 'species', 'NCBITAXON_9483')\n",
    "# metadata_marmoset.insert(4, 'species__ontology_label', 'Callithrix jacchus')\n",
    "# metadata_marmoset.insert(5, 'disease', 'PATO_0000461')\n",
    "# metadata_marmoset.insert(6, 'disease__ontology_label', 'normal')\n",
    "# metadata_marmoset.insert(7, 'organ', 'UBERON_0000966')\n",
    "# metadata_marmoset.insert(8, 'organ__ontology_label', 'retina')\n",
    "# metadata_marmoset.insert(9, 'library_preparation_protocol', 'EFO_0009899')\n",
    "# metadata_marmoset.insert(10, 'library_preparation_protocol__ontology_label', \"10X 3' v2\")\n",
    "# metadata_marmoset.insert(11, 'sex', 'unknown')\n",
    "\n",
    "# group_row = pd.Series(['TYPE'] + ['group'] * (len(metadata_marmoset.columns) - 1), index=metadata_marmoset.columns)\n",
    "# # Concatenate the 'group' row with the existing DataFrame\n",
    "# metadata_marmoset = pd.concat([group_row.to_frame().T, metadata_marmoset], ignore_index=True)\n",
    "\n",
    "# metadata_marmoset = metadata_marmoset.iloc[:, :12]\n",
    "\n",
    "\n",
    "\n",
    "# metadata_mouse.insert(1, 'biosample_id', 'P56_Batch1')\n",
    "# metadata_mouse.insert(2, 'donor_id', 'P56_Batch1')\n",
    "\n",
    "\n",
    "# metadata_mouse.insert(1, 'biosample_id', metadata_mouse['orig.file'])\n",
    "# metadata_mouse.insert(2, 'donor_id', metadata_mouse['animal'])\n",
    "\n",
    "# metadata_mouse.insert(3, 'species', 'NCBITaxon_10090')\n",
    "# metadata_mouse.insert(4, 'species__ontology_label', 'Mus musculus')\n",
    "# metadata_mouse.insert(5, 'disease', 'PATO_0000461')\n",
    "# metadata_mouse.insert(6, 'disease__ontology_label', 'normal')\n",
    "# metadata_mouse.insert(7, 'organ', 'UBERON_0000966')\n",
    "# metadata_mouse.insert(8, 'organ__ontology_label', 'retina')\n",
    "# metadata_mouse.insert(9, 'library_preparation_protocol', 'EFO_0009899')\n",
    "# metadata_mouse.insert(10, 'library_preparation_protocol__ontology_label', \"10X 3' v2\")\n",
    "# metadata_mouse.insert(11, 'sex', 'unknown')\n",
    "\n",
    "# group_row = pd.Series(['TYPE'] + ['group'] * (len(metadata_mouse.columns) - 1), index=metadata_mouse.columns)\n",
    "# # Concatenate the 'group' row with the existing DataFrame\n",
    "# metadata_mouse = pd.concat([group_row.to_frame().T, metadata_mouse], ignore_index=True)\n",
    "\n",
    "# metadata_mouse = metadata_mouse.iloc[:, :12]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
